### ISSUE TITLE
Can `AnalysisContext.getAnalysisOptionsForFile` be removed in favor of access from `ResolvedUnit`?

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#54667]
To resolve before publishing 7.0.0.

**Comment by pq** [comment]
@scheglov: were you proposing `ResolvedUnitResult`? Thoughts for cases (like `handleFileResult`) where we're only sure to have an `AnalysisResultWithErrors`?

**Comment by pq** [comment]
Chatting with Brian, I think we found that `FileResult` might be a good home.

@scheglov: thoughts?

**Comment by scheglov** [comment]
Yes, it is just another property of an analyzed file, so putting it into `FileResult` seems right.

**Comment by bwilkerson** [comment]
I'm convinced that we want to add this to `FileResult`. In particular, if we're not going to be rebuilding the analysis contexts when options are changed, then access to the options needs to go through the session in order to preserve consistency of results.

This does mean that when analysis options files are added, removed, or modified, a new analysis session needs to be created in order to reflect the fact that new analysis results might not be consistent with older results.

**Comment by scheglov** [comment]
Once an `AnalysisContextCollection` instance is created, we don't change anything about its analysis contexts configuration when configuration files like `analysis_options.yaml` change. DAS disposes the whole collection, and builds a new one.

**Comment by bwilkerson** [comment]
Today, yes. But if the presence of an options files doesn't create a new context then we wouldn't necessarily need to rebuild everything from scratch when one changes. By moving this API from `AnalysisContext` to `AnalysisSession` we guarantee that we won't be blocked if we later want to make that optimization.

**Comment by scheglov** [comment]
I agree.

**Comment by pq** [comment]
Ok, so I've hit some snags in https://dart-review.googlesource.com/c/sdk/+/347381. Let's regroup on Monday.

**Comment by srawlins** [comment]
This issue needs weekly updates, as it is P1. Any updates, @pq?
